src/tflite_cpu.cpp

```diff
@@ -0,0 +1,756 @@
+//
+// The following optional parameters are honored:
+//   -- `num_threads`: Sets the number of threads to be used, where applicable.
+//
+//   -- `label_path`:  An absolute filesystem path to a .txt file of the model's category labels.
+//
+// Any additional configuration fields are ignored.
+class MLModelServiceTFLite : public vsdk::MLModelService,
+                             public vsdk::Stoppable,
+                             public vsdk::Reconfigurable {
+    class write_to_tflite_tensor_visitor_;
+
+   public:
+    explicit MLModelServiceTFLite(vsdk::Dependencies dependencies,
+                                  vsdk::ResourceConfig configuration)
+        : MLModelService(configuration.name()),
+          state_(reconfigure_(std::move(dependencies), std::move(configuration))) {}
+
+    ~MLModelServiceTFLite() final {
+        // All invocations arrive via gRPC, so we know we are idle
+        // here. It should be safe to tear down all state
+        // automatically without needing to wait for anything more to
+        // drain.
+    }
+
+    void stop(const vsdk::ProtoStruct& extra) noexcept final {
+        return stop();
+    }
+
+    /// @brief Stops the MLModelServiceTFLite from running.
+    void stop() noexcept {
+        using std::swap;
+        try {
+            const std::lock_guard<std::mutex> lock(state_lock_);
+            if (!stopped_) {
+                stopped_ = true;
+                std::shared_ptr<state> state;
+                swap(state_, state);
+                state_ready_.notify_all();
+            }
+        } catch (...) {
+        }
+    }
+
+    void reconfigure(const vsdk::Dependencies& dependencies,
+                     const vsdk::ResourceConfig& configuration) final try {
+        // Care needs to be taken during reconfiguration. The
+        // framework does not offer protection against invocation
+        // during reconfiguration. Keep all state in a shared_ptr
+        // managed block, and allow client invocations to act against
+        // current state while a new configuration is built, then swap
+        // in the new state. State which is in use by existing
+        // invocations will remain valid until the clients drain. If
+        // reconfiguration fails, the component will `stop`.
+
+        // Swap out the state_ member with nullptr. Existing
+        // invocations will continue to operate against the state they
+        // hold, and new invocations will block on the state becoming
+        // populated.
+        using std::swap;
+        std::shared_ptr<state> state;
+        {
+            // Wait until we have a state in play, then take
+            // ownership, so that we don't race with other
+            // reconfigurations and so other invocations wait on a new
+            // state.
+            std::unique_lock<std::mutex> lock(state_lock_);
+            state_ready_.wait(lock, [this]() { return (state_ != nullptr) && !stopped_; });
```

Comment:
If this component ever becomes stopped, I think this will wait forever. and if the component was stopped because we're shutting the module down, I think the whole thing might hang waiting for this thread to exit. I like the call to `check_stopped_inlock()` on the next line; I think we should remove the `&& !stopped_` from this line.

_edit: if you get rid of the condition variable entirely (see other comment), this is a moot point._

Reply:
I'll ask @acmorrow to comment, as he originally wrote this code

Reply:
Yeah, looks wrong. The whole thing can go away if we eliminate the state swapping, which I will do more or less immediately after this lands (I can work on it this afternoon).